
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Custom Types &#8212; Common Workflow Language User Guide 0.1 documentation</title>
    
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="../_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <link rel="shortcut icon" href="../_static/favicon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Specifying Software Requirements" href="../20-software-requirements/index.html" />
    <link rel="prev" title="Metadata and Authorship" href="../17-metadata/index.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    
      
      <link rel="icon" sizes="16x16" href="../_static/images/favicons/cwl/favicon-16x16.png">
      
    
      
      <link rel="icon" sizes="32x32" href="../_static/images/favicons/cwl/favicon-32x32.png">
      
    
      
      <link rel="icon" sizes="96x96" href="../_static/images/favicons/cwl/favicon-96x96.png">
      
    
      
      <link rel="icon" sizes="128x128" href="../_static/images/favicons/cwl/favicon-128.png">
      
    
      
      <link rel="icon" sizes="196x196" href="../_static/images/favicons/cwl/favicon-196x196.png">
      
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="60">
    
    <div class="container-fluid" id="banner"></div>

    
    <nav class="navbar navbar-light navbar-expand-lg bg-light fixed-top bd-navbar" id="navbar-main"><div class="container-xl">

  <div id="navbar-start">
    
    

<a class="navbar-brand" href="../index.html">
  <img src="../_static/CWL-Logo-HD-cropped2.png" class="logo" alt="logo">
</a>


    
  </div>

  <button class="navbar-toggler" type="button" data-toggle="collapse" data-target="#navbar-collapsible" aria-controls="navbar-collapsible" aria-expanded="false" aria-label="Toggle navigation">
    <span class="navbar-toggler-icon"></span>
  </button>

  
  <div id="navbar-collapsible" class="col-lg-9 collapse navbar-collapse">
    <div id="navbar-center" class="mr-auto">
      
      <div class="navbar-center-item">
        <ul id="navbar-main-elements" class="navbar-nav">
    <li class="toctree-l1 nav-item">
 <a class="reference internal nav-link" href="../CODE_OF_CONDUCT.html">
  Code of Conduct
 </a>
</li>

<li class="toctree-l1 nav-item">
 <a class="reference internal nav-link" href="../setup.html">
  Setup
 </a>
</li>

<li class="toctree-l1 current active nav-item">
 <a class="reference internal nav-link" href="../episodes.html">
  Episodes
 </a>
</li>

<li class="toctree-l1 nav-item">
 <a class="reference internal nav-link" href="../extras.html">
  Extras
 </a>
</li>

<li class="toctree-l1 nav-item">
 <a class="reference internal nav-link" href="../LICENSE.html">
  Licenses
 </a>
</li>

    
</ul>
      </div>
      
    </div>

    <div id="navbar-end">
      
      <div class="navbar-end-item">
        <ul id="navbar-icon-links" class="navbar-nav" aria-label="Icon Links">
        <li class="nav-item">
          <a class="nav-link" href="https://github.com/common-workflow-language/user_guide/" rel="noopener" target="_blank" title="GitHub"><span><i class="fab fa-github-square"></i></span>
            <label class="sr-only">GitHub</label></a>
        </li>
      </ul>
      </div>
      
    </div>
  </div>
</div>
    </nav>
    

    <div class="container-xl">
      <div class="row">
          
            
            <!-- Only show if we have sidebars configured, else just a small margin  -->
            <div class="col-12 col-md-3 bd-sidebar">
              <div class="sidebar-start-items"><form class="bd-search d-flex align-items-center" action="../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search the docs ..." aria-label="Search the docs ..." autocomplete="off" >
</form><nav class="bd-links d-none d-md-block" id="bd-docs-nav" aria-label="Main navigation">
  <div class="bd-toc-item active">
    <ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../01-introduction/index.html">
   Introduction
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../02-1st-example/index.html">
   First Example
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../03-input/index.html">
   Essential Input Parameters
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../04-output/index.html">
   Returning Output Files
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../05-stdout/index.html">
   Capturing Standard Output
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../06-params/index.html">
   Parameter References
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../07-containers/index.html">
   Running Tools Inside Docker
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../08-arguments/index.html">
   Additional Arguments and Parameters
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../09-array-inputs/index.html">
   Array Inputs
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../10-array-outputs/index.html">
   Array Outputs
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../11-records/index.html">
   Advanced Inputs
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../12-env/index.html">
   Environment Variables
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../13-expressions/index.html">
   JavaScript Expressions
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../14-runtime/index.html">
   Creating Files at Runtime
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../15-staging/index.html">
   Staging Input Files
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../16-file-formats/index.html">
   File Formats
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../17-metadata/index.html">
   Metadata and Authorship
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   Custom Types
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../20-software-requirements/index.html">
   Specifying Software Requirements
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../21-1st-workflow/index.html">
   Writing Workflows
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../22-nested-workflows/index.html">
   Nested Workflows
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../23-scatter-workflow/index.html">
   Scattering Workflows
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../24_conditional-workflow/index.html">
   Conditional workflows
  </a>
 </li>
</ul>

  </div>
</nav>
              </div>
              <div class="sidebar-end-items">
              </div>
            </div>
            
          

          
          <div class="d-none d-xl-block col-xl-2 bd-toc">
            
              
              <div class="toc-item">
                

<nav id="bd-toc-nav">
    
</nav>
              </div>
              
              <div class="toc-item">
                

<div class="tocsection editthispage">
    <a href="https://github.com/common-workflow-language/user_guide/edit/gh-pages/19-custom-types/index.md">
        <i class="fas fa-pencil-alt"></i> Edit this page
    </a>
</div>

              </div>
              
            
          </div>
          

          
          
            
          
          <main class="col-12 col-md-9 col-xl-7 py-md-5 pl-md-5 pr-md-4 bd-content" role="main">
              
              <div>
                
  <section id="custom-types">
<h1>Custom Types<a class="headerlink" href="#custom-types" title="Permalink to this headline">#</a></h1>
<p>Sometimes you may want to write your own custom types for use and reuse in CWL
descriptions. Use of such custom types can reduce redundancy between multiple
descriptions that all use the same type, and also allow for additional
customisation/configuration of a tool/analysis without the need to fiddle with
the CWL description directly.</p>
<p>The example below is a CWL description of the <a class="reference external" href="http://biom-format.org/">biom convert format</a> tool for
converting a standard biom table file to hd5 format.</p>
<p><em>custom-types.cwl</em></p>
<div class="highlight-cwl notranslate"><div class="highlight"><pre><span></span><span class="c1">#!/usr/bin/env cwl-runner </span><span class="w"></span>
<span class="nt">cwlVersion</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">v1.0</span><span class="w"></span>
<span class="nt">class</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">CommandLineTool</span><span class="w"></span>

<span class="nt">requirements</span><span class="p">:</span><span class="w"></span>
<span class="w">  </span><span class="nt">InlineJavascriptRequirement</span><span class="p">:</span><span class="w"> </span><span class="p p-Indicator">{}</span><span class="w"></span>
<span class="w">  </span><span class="nt">ResourceRequirement</span><span class="p">:</span><span class="w"></span>
<span class="w">    </span><span class="nt">coresMax</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">1</span><span class="w"></span>
<span class="w">    </span><span class="nt">ramMin</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">100</span><span class="w">  </span><span class="c1"># just a default, could be lowered</span><span class="w"></span>
<span class="w">  </span><span class="nt">SchemaDefRequirement</span><span class="p">:</span><span class="w"></span>
<span class="w">    </span><span class="nt">types</span><span class="p">:</span><span class="w"></span>
<span class="w">      </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="nt">$import</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">biom-convert-table.yaml</span><span class="w"></span>

<span class="nt">hints</span><span class="p">:</span><span class="w"></span>
<span class="w">  </span><span class="nt">DockerRequirement</span><span class="p">:</span><span class="w"></span>
<span class="w">    </span><span class="nt">dockerPull</span><span class="p">:</span><span class="w"> </span><span class="s">&#39;quay.io/biocontainers/biom-format:2.1.6--py27_0&#39;</span><span class="w"></span>
<span class="w">  </span><span class="nt">SoftwareRequirement</span><span class="p">:</span><span class="w"></span>
<span class="w">    </span><span class="nt">packages</span><span class="p">:</span><span class="w"></span>
<span class="w">      </span><span class="nt">biom-format</span><span class="p">:</span><span class="w"></span>
<span class="w">        </span><span class="nt">specs</span><span class="p">:</span><span class="w"> </span><span class="p p-Indicator">[</span><span class="w"> </span><span class="s">&quot;https://doi.org/10.1186/2047-217X-1-7&quot;</span><span class="w"> </span><span class="p p-Indicator">]</span><span class="w"></span>
<span class="w">        </span><span class="nt">version</span><span class="p">:</span><span class="w"> </span><span class="p p-Indicator">[</span><span class="w"> </span><span class="s">&quot;2.1.6&quot;</span><span class="w"> </span><span class="p p-Indicator">]</span><span class="w"></span>

<span class="nt">inputs</span><span class="p">:</span><span class="w"></span>
<span class="w">  </span><span class="nt">biom</span><span class="p">:</span><span class="w"></span>
<span class="w">    </span><span class="nt">type</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">File</span><span class="w"></span>
<span class="w">    </span><span class="nt">format</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">edam:format_3746</span><span class="w">  </span><span class="c1"># BIOM</span><span class="w"></span>
<span class="w">    </span><span class="nt">inputBinding</span><span class="p">:</span><span class="w"></span>
<span class="w">      </span><span class="nt">prefix</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">--input-fp</span><span class="w"></span>
<span class="w">  </span><span class="nt">table_type</span><span class="p">:</span><span class="w"></span>
<span class="w">    </span><span class="nt">type</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">biom-convert-table.yaml#table_type</span><span class="w"></span>
<span class="w">    </span><span class="nt">inputBinding</span><span class="p">:</span><span class="w"></span>
<span class="w">      </span><span class="nt">prefix</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">--table-type</span><span class="w"></span>

<span class="w">  </span><span class="nt">header_key</span><span class="p">:</span><span class="w"></span>
<span class="w">    </span><span class="nt">type</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">string?</span><span class="w"></span>
<span class="w">    </span><span class="nt">doc</span><span class="p">:</span><span class="w"> </span><span class="p p-Indicator">|</span><span class="w"></span>
<span class="w">      </span><span class="no">The observation metadata to include from the input BIOM table file when</span><span class="w"></span>
<span class="w">      </span><span class="no">creating a tsv table file. By default no observation metadata will be</span><span class="w"></span>
<span class="w">      </span><span class="no">included.</span><span class="w"></span>
<span class="w">    </span><span class="nt">inputBinding</span><span class="p">:</span><span class="w"></span>
<span class="w">      </span><span class="nt">prefix</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">--header-key</span><span class="w"></span>

<span class="nt">baseCommand</span><span class="p">:</span><span class="w"> </span><span class="p p-Indicator">[</span><span class="w"> </span><span class="nv">biom</span><span class="p p-Indicator">,</span><span class="w"> </span><span class="nv">convert</span><span class="w"> </span><span class="p p-Indicator">]</span><span class="w"></span>

<span class="nt">arguments</span><span class="p">:</span><span class="w"></span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="nt">valueFrom</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">$(inputs.biom.nameroot).hdf5</span><span class="w">  </span>
<span class="w">    </span><span class="nt">prefix</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">--output-fp</span><span class="w"></span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">--to-hdf5</span><span class="w"></span>

<span class="nt">outputs</span><span class="p">:</span><span class="w"></span>
<span class="w">  </span><span class="nt">result</span><span class="p">:</span><span class="w"></span>
<span class="w">    </span><span class="nt">type</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">File</span><span class="w"></span>
<span class="w">    </span><span class="nt">outputBinding</span><span class="p">:</span><span class="w"> </span><span class="p p-Indicator">{</span><span class="nt"> glob</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;$(inputs.biom.nameroot)*&quot;</span><span class="w"> </span><span class="p p-Indicator">}</span><span class="w"></span>

<span class="nt">$namespaces</span><span class="p">:</span><span class="w"></span>
<span class="w">  </span><span class="nt">edam</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">http://edamontology.org/</span><span class="w"></span>
<span class="w">  </span><span class="nt">s</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">https://schema.org/</span><span class="w"></span>

<span class="nt">$schemas</span><span class="p">:</span><span class="w"></span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">http://edamontology.org/EDAM_1.16.owl</span><span class="w"></span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">https://schema.org/version/latest/schemaorg-current-http.rdf</span><span class="w"></span>

<span class="l l-Scalar l-Scalar-Plain">s:license</span><span class="p p-Indicator">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">https://spdx.org/licenses/Apache-2.0</span><span class="w"></span>
<span class="l l-Scalar l-Scalar-Plain">s:copyrightHolder</span><span class="p p-Indicator">:</span><span class="w"> </span><span class="s">&quot;EMBL</span><span class="nv"> </span><span class="s">-</span><span class="nv"> </span><span class="s">European</span><span class="nv"> </span><span class="s">Bioinformatics</span><span class="nv"> </span><span class="s">Institute&quot;</span><span class="w"></span>
</pre></div>
</div>
<p><em>custom-types.yml</em></p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">biom</span><span class="p">:</span><span class="w"></span>
<span class="w">    </span><span class="nt">class</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">File</span><span class="w"></span>
<span class="w">    </span><span class="nt">format</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">http://edamontology.org/format_3746</span><span class="w"></span>
<span class="w">    </span><span class="nt">path</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">rich_sparse_otu_table.biom</span><span class="w"></span>
<span class="nt">table_type</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">OTU table</span><span class="w"></span>
</pre></div>
</div>
<p><em><strong>Note:</strong></em> To follow the example below, you need to download the example input file, <em>rich_sparse_otu_table.biom</em>. The file is available from <a class="reference external" href="https://raw.githubusercontent.com/common-workflow-language/user_guide/gh-pages/_includes/cwl/19-custom-types/rich_sparse_otu_table.biom">https://raw.githubusercontent.com/common-workflow-language/user_guide/gh-pages/_includes/cwl/19-custom-types/rich_sparse_otu_table.biom</a> and can be downloaded e.g. via <code class="docutils literal notranslate"><span class="pre">wget</span></code>:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>wget https://raw.githubusercontent.com/common-workflow-language/user_guide/gh-pages/_includes/cwl/19-custom-types/rich_sparse_otu_table.biom
</pre></div>
</div>
<p>On line 29, in <code class="docutils literal notranslate"><span class="pre">inputs:table_type</span></code>, a list of allowable table options to be used in the
table conversion are imported as a custom object:</p>
<div class="highlight-cwl notranslate"><div class="highlight"><pre><span></span><span class="nt">inputs</span><span class="p">:</span><span class="w"></span>
<span class="w">  </span><span class="nt">biom</span><span class="p">:</span><span class="w"></span>
<span class="w">    </span><span class="nt">type</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">File</span><span class="w"></span>
<span class="w">    </span><span class="nt">format</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">edam:format_3746</span><span class="w">  </span><span class="c1"># BIOM</span><span class="w"></span>
<span class="w">    </span><span class="nt">inputBinding</span><span class="p">:</span><span class="w"></span>
<span class="w">      </span><span class="nt">prefix</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">--input-fp</span><span class="w"></span>
<span class="w">  </span><span class="nt">table_type</span><span class="p">:</span><span class="w"></span>
<span class="w">    </span><span class="nt">type</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">biom-convert-table.yaml#table_type</span><span class="w"></span>
<span class="w">    </span><span class="nt">inputBinding</span><span class="p">:</span><span class="w"></span>
<span class="w">      </span><span class="nt">prefix</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">--table-type</span><span class="w"></span>
</pre></div>
</div>
<p>The reference to a custom type is a combination of the name of the file in which
the object is defined (<code class="docutils literal notranslate"><span class="pre">biom-convert-table.yaml</span></code>) and the name of the object
within that file (<code class="docutils literal notranslate"><span class="pre">table_type</span></code>) that defines the custom type. In this case the <code class="docutils literal notranslate"><span class="pre">symbols</span></code>
array from the imported <code class="docutils literal notranslate"><span class="pre">biom-convert-table.yaml</span></code> file define the allowable table options.
For example, in <code class="docutils literal notranslate"><span class="pre">custom-types.yml</span></code>, we pass <code class="docutils literal notranslate"><span class="pre">OTU</span> <span class="pre">table</span></code> as an <code class="docutils literal notranslate"><span class="pre">input</span></code> that
tells the tool to create an OTU table in hd5 format.</p>
<p>The contents of the YAML file describing the custom type are given below:</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">type</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">enum</span><span class="w"></span>
<span class="nt">name</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">table_type</span><span class="w"></span>
<span class="nt">label</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">The type of the table to produce</span><span class="w"></span>
<span class="nt">symbols</span><span class="p">:</span><span class="w"></span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">OTU table</span><span class="w"></span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">Pathway table</span><span class="w"></span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">Function table</span><span class="w"></span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">Ortholog table</span><span class="w"></span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">Gene table</span><span class="w"></span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">Metabolite table</span><span class="w"></span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">Taxon table</span><span class="w"></span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">Table</span><span class="w"></span>
</pre></div>
</div>
<p>In order for the custom type to be used in the CWL description, it must be
imported. Imports are described in <code class="docutils literal notranslate"><span class="pre">requirements:SchemaDefRequirement</span></code>, as
below in the example <code class="docutils literal notranslate"><span class="pre">custom-types.cwl</span></code> description:</p>
<div class="highlight-cwl notranslate"><div class="highlight"><pre><span></span><span class="nt">requirements</span><span class="p">:</span><span class="w"></span>
<span class="w">  </span><span class="nt">InlineJavascriptRequirement</span><span class="p">:</span><span class="w"> </span><span class="p p-Indicator">{}</span><span class="w"></span>
<span class="w">  </span><span class="nt">ResourceRequirement</span><span class="p">:</span><span class="w"></span>
<span class="w">    </span><span class="nt">coresMax</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">1</span><span class="w"></span>
<span class="w">    </span><span class="nt">ramMin</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">100</span><span class="w"></span>
<span class="w">  </span><span class="nt">SchemaDefRequirement</span><span class="p">:</span><span class="w"></span>
<span class="w">    </span><span class="nt">types</span><span class="p">:</span><span class="w"></span>
<span class="w">      </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="nt">$import</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">biom-convert-table.yaml</span><span class="w"></span>
</pre></div>
</div>
<p>Note also that the author of this CWL description has also included
<code class="docutils literal notranslate"><span class="pre">ResourceRequirement</span></code>s, specifying the minimum amount of RAM and number of cores
required for the tool to run successfully, as well as details of the version of
the software that the description was written for and other useful metadata.
These features are discussed further in other chapters of this user guide.</p>
</section>


              </div>
              
              
              <!-- Previous / next buttons -->
<div class='prev-next-area'>
    <a class='left-prev' id="prev-link" href="../17-metadata/index.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Metadata and Authorship</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="../20-software-requirements/index.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Specifying Software Requirements</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
              
          </main>
          

      </div>
    </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>

<div class="container">
  <div class="row">
    <div class="col d-flex flex-row-reverse">
      <ul class="nav">
        <li class="nav-item">
          <a class="nav-link" href="https://github.com/common-workflow-language/user_guide/blob/gh-pages/CONTRIBUTING.md">Contributing</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="https://github.com/common-workflow-language/user_guide/">Source</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="https://github.com/common-workflow-language/user_guide/blob/gh-pages/CITATION">Cite</a>
        </li>
      </ul>
    </div>
  </div>
</div>

<footer class="footer mt-5 mt-md-0">
  <div class="container">
    
    <div class="footer-item">
      <p class="copyright">
    &copy; Copyright 2013, CWL Project Team.<br>
</p>
    </div>
    
  </div>
</footer>

  </body>
</html>